docs/source/more_types.rst

```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+First, if multiple variants match due to an argument being of type
+``Any``, mypy will make the inferred type also be ``Any``:
+
+.. code-block:: python
+
+    dynamic_var: Any = some_dynamic_function()
+
+    # output2 is of type 'Any'
+    output2 = summarize(dynamic_var)
+
+Second, if multiple variants match due to one or more of the arguments
+being a union, mypy will make the inferred type be the union of the 
+matching variant returns:
+
+.. code-block:: python
+
+    some_list: Union[List[str], List[int]]
+
+    # output3 is of type 'Union[str, int]'
+    output3 = summarize(some_list)
+
+.. note::
+
+   Due to the "pick the first match" rule, changing the order of your
+   overload variants can change how mypy type checks your program.
+
+   To minimize potential issues, we recommend ordering your variants
+   from most to least specific. Your implementation should also
+   perform runtime checks (e.g. ``isinstance`` checks) in the same
+   order as the listed variants.
+
+   If your variants have no inherent relationship to each other, they
+   can naturally be listed in any arbitrary order.
+
+Type checking the variants
+--------------------------
+
+Although the "pick the first match" algorithm works well in many cases, it can
+sometimes make mypy infer the incorrect type. For example, consider the following
+overload definition:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    @overload
+    def unsafe_func(x: int) -> int: ...
+
+    @overload
+    def unsafe_func(x: object) -> str: ...
+
+    def unsafe_func(x: object) -> Union[int, str]:
+        if isinstance(x, int):
+            return x
+        else:
+            return "Unsafe!"
+
+On the surface, this function definition appears to be fine. However, it will
+result in a discrepency between the inferred type and the actual runtime type
+when we try using it like so:
+
+.. code-block:: python
+
+    def should_return_str(x: object) -> str:
+        return unsafe_func(x)
+
+    bad_var = should_return_str(42)
+
+If we examine just the annotated types, it seems as if ``bad_var`` ought to be of
+type ``str``. But if we examine the runtime behavior, ``bad_var`` will actually be
+of type ``int``!
+
+To prevent these kinds of issues, mypy will detect and prohibit inherently unsafely
+overlapping overloads on a best-effort basis. Two variants are considered unsafely
+overlapping when both of the following are true:
+
+1. All of the arguments of the first variant are compatible with the second.
+2. The return type of the first variant is *not* compatible with (e.g. is not a
+   subtype of) the second.
+
+So in this example, the ``int`` argument in the first variant is a subtype of
+the ``object`` argument in the second, yet the ``int`` return type is a subtype of
+``str``. Both conditions are true, so mypy will correctly flag ``unsafe_func`` as
+being unsafe.
+
+However, it is unfortunately impossible to detect *all* unsafe overloads without
```

Comment:
All this long part about multiple inheritance is unnecessary I think. It can be replaced with a single short note at the end of this section like: "Note that mypy intentionally allows unsafe overloads related to multiple inheritance and empty containers. This way mypy avoids lots of false positives (that would otherwise make overloads impractical) by having some rare false negatives."

This is really a discussion that belongs to the PEP, not the docs. Remember that you write this mostly for people who have never heard about overloads, while currently it looks like you are talking to a CS prof: "Please forgive us for being so unsafe!" :-)

Reply:
Mm, I'm ok with just leaving a short snippet like that if we're ok with also linking to the PEP and telling the reader to read that if they want more details.

But I was under the impression that we wanted the docs to be more self-contained? In that case, I think we really do need at least one example of what an unsafe yet allowed overload could look like. I don't think a reader would really understand what we mean if with just that explanation .

One option I could try is re-using the `summarize` example instead -- that would let me cut out all of the setup related to multiple-inheritance which I think would help cut down the text by a decent amount. I still want to keep the gist of recommendations at the end and some of the exposition though. Empty collections are relatively common and some people do like to use mixins -- I think it's worth giving people a more detailed heads up.

Reply:
> But I was under the impression that we wanted the docs to be more self-contained?

Self-contained doesn't mean complete, in particular here too much attention is devoted to motivation why we allow this. You could try re-using `summarize` if this will help cutting the size of this significantly.